### PR TITLE
Hotfix/global filters

### PR DESCRIPTION
**Pull request by Ravikant96**
Description:
Describe your pull request and attach links to any issues that may be related.

### CODE REVIEW

File: client/index.js [hunk 1]
```diff
@@ -1389,18 +1393,40 @@ router.get('/settings/:tab?/:id?', API.serve(class extends HTMLAPI {
...
+						<label>
+							<span>Type</span>
+							<select name="type">
+								<option value="text">Text</option>
+								<option value="number">Number</option>
+								<option value="date">Date</option>
+								<option value="month">Month</option>
+								<option value="hidden">Hidden</option>
+								<option value="column">Column</option>
+								<option value="datetime">Datetime</option>
```
Comment (achshar) [inline client/index.js]: `MetaData.inputTypes`

File: client/index.js [hunk 2]
```diff
@@ -1347,23 +1347,27 @@ router.get('/settings/:tab?/:id?', API.serve(class extends HTMLAPI {
...
 					<header class="toolbar">
-						<button id="add-datset"><i class="fa fa-plus"></i> Add New Dataset</button>
+						<button id="add-global-filter"><i class="fa fa-plus"></i> Add New Global Filter</button>
 					</header>
 
 					<table class="block">
 						<thead>
 							<tr>
 								<th>ID</th>
+								<th>Account ID</th>
```
Comment (achshar) [inline client/index.js]: Account id not needed.

File: client/js/dashboard.js [hunk 3]
```diff
@@ -1228,7 +1228,7 @@ class DashboardGlobalFilters extends DataSourceFilters {
 
 			for(const filter of visualization.filters.values()) {
 
-				if(globalFilters.has(filter.placeholder) || ['hidden', 'daterange'].includes(filter.type))
+				if(!Array.from(MetaData.globalFilters).some(a => a[1].placeholder == filter.placeholder))
```
Comment (achshar) [inline client/js/dashboard.js]: `f.placeholders` is an array.

File: client/js/settings.js [hunk 4]
```diff
@@ -1174,4 +1049,156 @@ class SettingsCategory {
...
+	static async insert(e, globalFilters) {
+
+		e.preventDefault();
+
+		const options = {
+			method: 'POST',
+			form: new FormData(globalFilters.form),
+		}
+
+		const response = await API.call('global_filters/insert', {dataset: globalFilters.multiselect.value}, options);
```
Comment (achshar) [inline client/js/settings.js]: Hyphen

File: client/js/settings.js [hunk 5]
```diff
@@ -1174,4 +1049,156 @@ class SettingsCategory {
...
+		this.multiselect = new MultiSelect({datalist, multiple: false});
+		this.multiselect.value = this.dataset;
+
+		this.globalFilters.form.querySelector('.datasets').appendChild(this.multiselect.container);
+
+		this.globalFilters.container.querySelector('#global-filters-form h1').textContent = 'Edit ' + this.name;
+
+		this.globalFilters.form.reset();
+
+		this.globalFilters.form.name.value = this.name;
```
Comment (achshar) [inline client/js/settings.js]: Use a `for...in` loop.

File: client/js/settings.js [hunk 6]
```diff
@@ -1174,4 +1049,156 @@ class SettingsCategory {
...
+	async update(e) {
+
+		e.preventDefault();
+
+		const parameter = {
+			id: this.id,
+			dataset: this.multiselect.value,
+		}
+
+		const options = {
```
Comment (achshar) [inline client/js/settings.js]: Single const.

File: client/js/settings.js [hunk 7]
```diff
@@ -1174,4 +1049,156 @@ class SettingsCategory {
...
+			form: new FormData(this.globalFilters.form),
+		}
+
+		await API.call('global_filters/update', parameter, options);
+
+		await this.globalFilters.load();
+
+		this.globalFilters.list.get(this.id).edit();
+		await Sections.show('global-filters-form');
+		this.globalFilters.list.get(this.id).edit();
```
Comment (achshar) [inline client/js/settings.js]: Only needed once.

File: client/js/settings.js [hunk 8]
```diff
@@ -1174,4 +1049,156 @@ class SettingsCategory {
...
+
+	async delete() {
+
+		if (!confirm('Are you sure?'))
+			return;
+
+		const options = {
+			method: 'POST',
+		}
+		const parameter = {
```
Comment (achshar) [inline client/js/settings.js]: Single const.

File: server/www/global_filters.js [hunk 9]
```diff
@@ -0,0 +1,65 @@
...
+		);
+	}
+}
+
+exports.update = class extends API {
+	async update() {
+		this.user.privilege.needs('administrator');
+
+		const params = {
+			account_id: this.account.account_id,
```
Comment (achshar) [inline server/www/global_filters.js]: account_id will never update.

File: server/www/global_filters.js [hunk 10]
```diff
@@ -0,0 +1,65 @@
...
+			multiple: this.request.body.multiple,
+			type: this.request.body.type,
+			offset: this.request.body.offset || null,
+			dataset: this.request.body.dataset,
+		};
+
+		return await this.mysql.query(
+			`UPDATE tb_global_filters SET ? WHERE id = ?`,
+			[params, this.request.body.id],
+			'write'
```
Comment (achshar) [inline server/www/global_filters.js]: Add a check to make sure user can only update it's own account's filters.

File: server/www/global_filters.js [hunk 11]
```diff
@@ -0,0 +1,65 @@
...
+		);
+	}
+}
+
+exports.delete = class extends API {
+	async delete() {
+		this.user.privilege.needs('administrator');
+
+		return await this.mysql.query(
+			`DELETE FROM tb_global_filters WHERE id = ?`,
```
Comment (achshar) [inline server/www/global_filters.js]: Add a check to make sure user can only update it's own account's filters.

File: server/www/global_filters.js [hunk 12]
```diff
@@ -0,0 +1,65 @@
+const API = require('../utils/api');
+
+exports.list = class extends API {
```
Comment (achshar) [inline server/www/global_filters.js]: Single class.

File: client/index.js (hunk 1 above)
Comment (Ravikant96) [reply client/index.js]: `MetaData.filterTypes`

File: client/js/settings.js [hunk 13]
```diff
@@ -74,55 +74,65 @@ class SettingPage {
...
-		this.form = this.container.querySelector('section#datasets-form form');
+		this.container = this.page.querySelector('.global-filters-page');
+		this.form = this.container.querySelector('section#global-filters-form form');
 
-		for (const data of MetaData.categories.values()) {
-			this.form.category_id.insertAdjacentHTML('beforeend', `
-				<option value="${data.category_id}">${data.name}</option>
-			`);
-		}
+		const datasourceResponse = await API.call('reports/report/list');
```
Comment (achshar) [inline client/js/settings.js]: `DataSource.list`

**Pull request merged**